File: financial-manager-app/lib/features/categories/domain/models/category.dart
import '../../../transactions/domain/models/transaction_direction.dart';

/// Which transaction directions a category applies to (plan.md section
/// 11.7). [both] categories (none seeded by default) show up as a
/// suggestion regardless of whether the operation is an entrata or uscita.
enum CategoryDirectionScope {
  debit,
  credit,
  both;

  static CategoryDirectionScope fromApi(String value) => switch (value) {
    'DEBIT' => CategoryDirectionScope.debit,
    'CREDIT' => CategoryDirectionScope.credit,
    'BOTH' => CategoryDirectionScope.both,
    _ => throw ArgumentError('Unknown category direction scope: $value'),
  };

  String toApi() => switch (this) {
    CategoryDirectionScope.debit => 'DEBIT',
    CategoryDirectionScope.credit => 'CREDIT',
    CategoryDirectionScope.both => 'BOTH',
  };
}

/// Economic classification of a transaction (plan.md section 4.1, 11.7):
/// shared system categories (Casa, Stipendio, ...) plus each user's own
/// custom ones.
class Category {
  const Category({
    required this.id,
    required this.name,
    required this.directionScope,
    this.color,
    required this.isSystem,
    required this.sortOrder,
  });

  final String id;
  final String name;
  final CategoryDirectionScope directionScope;
  final String? color;
  final bool isSystem;
  final int sortOrder;

  /// Whether this category should be offered for an operation of the
  /// given [direction] (plan.md section 7.6: "modelli di uscita per
  /// uscite, di entrata per entrate" — the same filtering rule applies to
  /// categories).
  bool matches(TransactionDirection direction) {
    return switch (directionScope) {
      CategoryDirectionScope.both => true,
      CategoryDirectionScope.debit => direction == TransactionDirection.debit,
      CategoryDirectionScope.credit => direction == TransactionDirection.credit,
    };
  }

  static Category fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'] as String,
      name: json['name'] as String,
      directionScope: CategoryDirectionScope.fromApi(
        json['direction_scope'] as String,
      ),
      color: json['color'] as String?,
      isSystem: json['is_system'] as bool,
      sortOrder: json['sort_order'] as int,
    );
  }
}

File: financial-manager-backend/internal/categories/category.go
// Package categories implements economic classification of transactions
// (plan.md section 4.1, 11.7) — a shared set of system categories plus
// each user's own custom ones.
package categories

import (
	"strings"
	"time"

	"github.com/google/uuid"
)

const (
	ScopeDebit  = "DEBIT"
	ScopeCredit = "CREDIT"
	ScopeBoth   = "BOTH"
)

type Category struct {
	ID             uuid.UUID
	OwnerUserID    *uuid.UUID
	Name           string
	NameNormalized string
	DirectionScope string
	IconMediaID    *uuid.UUID
	Color          *string
	IsSystem       bool
	SortOrder      int
	CreatedAt      time.Time
	UpdatedAt      time.Time
	ArchivedAt     *time.Time
}

// NormalizeName mirrors plan.md section 4.4's title normalization: trim,
// compact internal spaces, case-insensitive comparison.
func NormalizeName(name string) string {
	return strings.ToLower(strings.Join(strings.Fields(strings.TrimSpace(name)), " "))
}

func IsValidScope(scope string) bool {
	return scope == ScopeDebit || scope == ScopeCredit || scope == ScopeBoth
}

// Matches reports whether the category can be used for a transaction with
// the given direction (plan.md section 7.6: "modelli di uscita per uscite,
// di entrata per entrate" — the same filtering rule applies to categories).
func (c Category) Matches(direction string) bool {
	return c.DirectionScope == ScopeBoth || c.DirectionScope == direction
}
